### PR TITLE
Flat nested objects to plain

### PR DESCRIPTION
**Pull request by AlexBelch**
Refs: [#113](https://github.com/metarhia/metautil/issues/113)    

- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fmt`)
- [x] description of changes is added in CHANGELOG.md
- [x] update .d.ts typings

### CODE REVIEW

File: lib/utilities.js [hunk 1]
```diff
@@ -329,6 +329,36 @@ const jsonParse = (buffer) => {
   }
 };
 
+const flatObject = (sourceObj, nameField) => {
+  const targetObj = {};
+
+  for (const [key, value] of Object.entries(sourceObj)) {
+    if (
+      typeof value !== 'object' ||
+      (typeof value === 'object' && Array.isArray(value))
```
Comment (dromix) [inline lib/utilities.js:338]: I suggest extracting it to function 
```suggestion
      !isObject(value)
```

Comment (AlexBelch) [reply lib/utilities.js]: @dromix @georgolden 
Done 

and now I have question 
need new func isObject export from utilities 
to use in another files ?

**Comment by georgolden** [comment]
Please add tests for multiple nested objects and try make it work recursively.
```js
metatests.test('Object: flatFields multiple nested objects', (test) => {
  const source = {
    name: { first: 'Andrew', second: 'Johnson' },
    old: true,
    avoid: [1, 2, 3],
    parent: { mother: { name: 'Eva' } , father: { name: 'Adam' } },
  };
  const expected = {
    nameFirst: 'Andrew',
    nameSecond: 'Johnson',
    old: true,
    avoid: [1, 2, 3],
    parentMotherName: 'Eva',
    parentFatherName: 'Adam',
  };

  const result = metautil.flatObject(source);

  test.strictSame(result, expected);
  test.end();
});
```

For field names it must be only first level checked

```js
metatests.test('Object: flatFields with key names', (test) => {
  const source = {
    name: { first: 'Andrew', second: 'Johnson' },
    old: true,
    parent: { mother: 'Eva', father: 'Adam' },
  };
  const expected = {
    nameFirst: 'Andrew',
    nameSecond: 'Johnson',
    old: true,
    parent: { mother: { name: 'Eva' } , father: { name: 'Adam' } },
  };

  const result = metautil.flatObject(source, 'name');

  test.strictSame(result, expected);
  test.end();
});
```

File: lib/utilities.js [hunk 2]
```diff
@@ -329,6 +332,33 @@ const jsonParse = (buffer) => {
...
+      continue;
+    }
+
+    for (const [key2, value2] of Object.entries(value)) {
+      const newKey = `${key}${toUpperCamel(key2)}`;
+
+      if (sourceObj[newKey] !== undefined)
+        throw new Error('flatObject: duplicate key');
+
+      targetObj[newKey] = value2;
```
Comment (dromix) [inline lib/utilities.js:355]: if here nested object you will copy a link
```javascript
  const source = {
    name: { first: 'Andrew', second: 'Johnson' },
    old: true,
    avoid: [1, 2, 3],
    parent: { mother: 'Eva', father: 'Adam', relatives: { stepBrother: 'Adam' } },
  };

 const expected = {
    nameFirst: 'Andrew', 
    nameSecond: 'Johnson' ,
    old: true,
    avoid: [1, 2, 3],
    parentMother: 'Eva', 
    parentFather: 'Adam', 
    parentRelatives: { stepBrother: 'Adam' },
  };

  source.parent.relatives.stepBrother = 'Tom';
  console.log(expected.parentRelatives.stepBrother); // 'Tom'
```
```suggestion
      targetObj[newKey] = isObject(value2) ? {...value2} : value2;
```
@AlexBelch

File: test/utilities.js [hunk 3]
```diff
@@ -432,3 +432,83 @@ metatests.test('Object: namespaceByPath', (test) => {
...
+  try {
+    metautil.flatObject(source);
+  } catch (err) {
+    test.strictSame(err.message, 'flatObject: duplicate key');
+  }
+
+  test.end();
+});
+
+metatests.test('Object: flatFields multiple nested objects', (test) => {
```
Comment (georgolden) [inline test/utilities.js:494]: We decided to not have such functionality now.

**Review by georgolden — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 ## [Unreleased][unreleased]
 
+- Add function `flatObject(sourceObj: object, nameField?: string): object`
```
Comment (tshemsedinov) [inline CHANGELOG.md:5]:
```suggestion
- Add function `flatObject(sourceObject: object, fieldNames: Array<string>): object`
```

File: lib/utilities.js [hunk 5]
```diff
@@ -329,6 +332,33 @@ const jsonParse = (buffer) => {
   }
 };
 
+const flatObject = (sourceObj, nameField) => {
```
[the remark this answers is not in the document]
Comment (tshemsedinov) [reply lib/utilities.js:335]: `fieldNames` is required

File: lib/utilities.js [hunk 6]
```diff
@@ -329,6 +332,33 @@ const jsonParse = (buffer) => {
...
+  const targetObj = {};
+
+  for (const [key, value] of Object.entries(sourceObj)) {
+    if (!isObject(value)) {
+      targetObj[key] = value;
+      continue;
+    }
+
+    if (nameField !== undefined && nameField !== key) {
```
Comment (tshemsedinov) [inline lib/utilities.js:344]: fieldNames is not optional

File: lib/utilities.js [hunk 7]
```diff
@@ -334,6 +337,33 @@ const jsonParse = (buffer) => {
...
+
+      if (sourceObject[newKey] !== undefined)
+        throw new Error('flatObject: duplicate key');
+
+      targetObj[newKey] = value2;
+    }
+  }
+
+  return targetObj;
+};
```
Comment (georgolden) [inline lib/utilities.js:365]: I suggest slightly naming improvement and default value usage for `fieldNames` argument
```js
const flatObject = (sourceObject, fieldNames = []) => {
  const target = {};

  for (const [key, value] of Object.entries(sourceObject)) {
    if (!isObject(value)) {
      target[key] = value;
      continue;
    }

    if (fieldNames.length > 0 && !fieldNames.includes(key)) {
      target[key] = { ...value };
      continue;
    }

    for (const [childKey, childValue] of Object.entries(value)) {
      const combinedKey = `${key}${toUpperCamel(childKey)}`;

      if (sourceObject[combinedKey])
        throw new Error(
          `Can not combine keys: key "${combinedKey}" already exists`,
        );

      target[combinedKey] = childValue;
    }
  }

  return target;
};
```

File: test/utilities.js [hunk 8]
```diff
@@ -447,3 +447,64 @@ metatests.test('Object: namespaceByPath', (test) => {
...
+    nameFirst: 'Andrew',
+    old: true,
+    parent: { mother: 'Eva', father: 'Adam' },
+  };
+
+  try {
+    metautil.flatObject(source, []);
+  } catch (err) {
+    test.strictSame(err.message, 'flatObject: duplicate key');
+  }
```
Comment (georgolden) [inline test/utilities.js:507]: I suggest to use `test.throws` from metatests
```js
  test.throws(
    () => metautil.flatObject(source),
    new Error('Can not combine keys: key "nameFirst" already exists'),
  );
```

File: test/utilities.js [hunk 9]
```diff
@@ -447,3 +447,64 @@ metatests.test('Object: namespaceByPath', (test) => {
...
+  const expected = {
+    nameFirst: 'Andrew',
+    nameSecond: 'Johnson',
+    old: true,
+    avoid: [1, 2, 3],
+    parentMother: 'Eva',
+    parentFather: 'Adam',
+  };
+
+  const result = metautil.flatObject(source, []);
```
Comment (georgolden) [inline test/utilities.js:467]: Don't need to pass an empty array as an argument. Use default value instead.

File: test/utilities.js [hunk 10]
```diff
@@ -447,3 +447,64 @@ metatests.test('Object: namespaceByPath', (test) => {
...
+metatests.test('Object: flatFields duplicate key', (test) => {
+  const source = {
+    name: { first: 'Andrew', second: 'Johnson' },
+    nameFirst: 'Andrew',
+    old: true,
+    parent: { mother: 'Eva', father: 'Adam' },
+  };
+
+  try {
+    metautil.flatObject(source, []);
```
Comment (georgolden) [inline test/utilities.js:504]: Don't need to pass an empty array as an argument. Use default value instead.

**Review by georgolden — CHANGES_REQUESTED** [review]

**Review by georgolden — APPROVED** [review]

File: lib/utilities.js [hunk 11]
```diff
@@ -356,6 +359,34 @@ const jsonParse = (buffer) => {
...
+      continue;
+    }
+
+    for (const [childKey, childValue] of Object.entries(value)) {
+      const combinedKey = `${key}${toUpperCamel(childKey)}`;
+
+      if (sourceObject[combinedKey] !== undefined)
+        throw new Error(
+          `Can not combine keys: key "${combinedKey}" already exists`,
+        );
```
Comment (tshemsedinov) [inline lib/utilities.js:382]:
```suggestion
      if (sourceObject[combinedKey] !== undefined) {
        const error = `Can not combine keys: key "${combinedKey}" already exists`;
        throw new Error(error);
      }
```

Comment (AlexBelch) [reply lib/utilities.js]: @tshemsedinov 
Done

File: lib/utilities.js [hunk 12]
```diff
@@ -87,6 +87,9 @@ const isFirstLower = (s) => !!s && inRange(s[0], 'a', 'z');
 
 const isFirstLetter = (s) => isFirstUpper(s) || isFirstLower(s);
 
+const isObject = (o) =>
+  typeof o === 'object' && o !== null && !Array.isArray(o);
+
```
Comment (tshemsedinov) [inline lib/utilities.js:92]:
```suggestion
const isHashObject = (o) =>
  typeof o === 'object' && o !== null && !Array.isArray(o);

```

File: lib/utilities.js [hunk 13]
```diff
@@ -356,6 +359,35 @@ const jsonParse = (buffer) => {
   }
 };
 
+const flatObject = (sourceObject, fieldNames = []) => {
+  const target = {};
+
+  for (const [key, value] of Object.entries(sourceObject)) {
+    if (!isObject(value)) {
```
Comment (tshemsedinov) [inline lib/utilities.js:366]:
```suggestion
    if (!isHashObject(value)) {
```

File: lib/utilities.js [hunk 14]
```diff
@@ -370,6 +402,7 @@ module.exports = {
   isFirstUpper,
   isFirstLower,
   isFirstLetter,
+  isObject,
```
Comment (tshemsedinov) [inline lib/utilities.js:405]:
```suggestion
  isHashObject,
```

File: metautil.d.ts [hunk 15]
```diff
@@ -40,6 +40,7 @@ export function toCamel(separator: string): (s: string) => string;
 export function spinalToCamel(s: string): string;
 export function snakeToCamel(s: string): string;
 export function isConstant(s: string): boolean;
+export function isObject(o: string | number | boolean | object): boolean;
```
Comment (tshemsedinov) [inline metautil.d.ts:43]:
```suggestion
export function isHashObject(o: string | number | boolean | object): boolean;
```

File: CHANGELOG.md [hunk 16]
```diff
@@ -7,6 +7,8 @@
 - New functions: `parseDay` and `parseMonth`
 - Fixed and `parseEvery` and `nextEvent`, more tests cases
 - Add year support to `Every` format
+- Add function `flatObject(sourceObject: object, fieldNames: Array<string>): object`
+- Add function `isObject(o: string | number | boolean | object ): boolean`
```
Comment (tshemsedinov) [inline CHANGELOG.md:11]:
```suggestion
- Add function `isHashObject(o: string | number | boolean | object): boolean`
```

**Review by tshemsedinov — APPROVED** [review]
LGTM

**Pull request merged**